File: bbike/app/api/v1/endpoints/bikes.py
from fastapi import APIRouter, HTTPException, Depends
from sqlalchemy.orm import Session
from sqlalchemy import func
from app.core.database import get_db
from app.models.snapshot import BikeSnapshot

router = APIRouter()

@router.get("/bikes")
async def get_bikes(db: Session = Depends(get_db)):
    # Get latest timestamp
    latest_time = db.query(func.max(BikeSnapshot.timestamp)).scalar()
    
    if not latest_time:
        raise HTTPException(status_code=503, detail="No bike data available yet")
    
    # Get all bikes from latest snapshot
    bikes = db.query(BikeSnapshot).filter(
        BikeSnapshot.timestamp == latest_time
    ).all()
    
    print(f"Total bikes: {len(bikes)}")
    
    return {
        "last_updated": latest_time.isoformat(),
        "data": [
            {
                "bike_id": b.bike_id,
                "lat": b.lat,
                "lon": b.lon,
                "is_reserved": b.is_reserved,
                "is_disabled": b.is_disabled,
                "vehicle_type_id": b.vehicle_type_id
            }
            for b in bikes
        ]
    }

@router.get("/bikes/available")
async def get_available_bikes(db: Session = Depends(get_db)):
    latest_time = db.query(func.max(BikeSnapshot.timestamp)).scalar()
    
    if not latest_time:
        raise HTTPException(status_code=503, detail="No bike data available yet")
    
    # Get available bikes only
    bikes = db.query(BikeSnapshot).filter(
        BikeSnapshot.timestamp == latest_time,
        BikeSnapshot.is_reserved == False,
        BikeSnapshot.is_disabled == False
    ).all()
    
    print(f"Available bikes: {len(bikes)}")
    
    return {
        "last_updated": latest_time.isoformat(),
        "available_bikes": [
            {
                "bike_id": b.bike_id,
                "lat": b.lat,
                "lon": b.lon,
                "vehicle_type_id": b.vehicle_type_id
            }
            for b in bikes
        ]
    }

@router.get("/ebikes")
async def get_ebikes(db: Session = Depends(get_db)):
    latest_time = db.query(func.max(BikeSnapshot.timestamp)).scalar()
    
    if not latest_time:
        raise HTTPException(status_code=503, detail="No bike data available yet")
    
    # Get ebikes only
    ebikes = db.query(BikeSnapshot).filter(
        BikeSnapshot.timestamp == latest_time,
        BikeSnapshot.vehicle_type_id == "bbe"
    ).all()
    
    return {
        "last_updated": latest_time.isoformat(),
        "ebikes": [
            {
                "bike_id": b.bike_id,
                "lat": b.lat,
                "lon": b.lon,
                "is_reserved": b.is_reserved,
                "is_disabled": b.is_disabled
            }
            for b in ebikes
        ]
    }

File: worker/src/routes/bikes.ts
import { Hono } from "hono";
import type { Env } from "../index";
import { resolveCity } from "../cities";

// Ported from bbike/app/api/v1/endpoints/bikes.py
// 503 only when there is no snapshot at all; filtered queries may return empty lists.

type BikeRow = {
  timestamp: string;
  bike_id: string;
  lat: number;
  lon: number;
  is_reserved: number;
  is_disabled: number;
  vehicle_type_id: string;
};

async function latestBikeTimestamp(db: D1Database, city: string): Promise<string | null> {
  return db
    .prepare("SELECT MAX(timestamp) AS ts FROM bike_snapshots WHERE city = ?")
    .bind(city)
    .first<string>("ts");
}

const bikes = new Hono<{ Bindings: Env }>();

bikes.get("/bikes", async (c) => {
  const city = resolveCity(c.req.query("city"));
  if (!city) return c.json({ detail: "Unknown city" }, 400);
  const ts = await latestBikeTimestamp(c.env.DB, city);
  if (!ts) return c.json({ detail: "No bike data available yet" }, 503);
  const { results } = await c.env.DB.prepare(
    "SELECT * FROM bike_snapshots WHERE city = ? AND timestamp = ?"
  )
    .bind(city, ts)
    .all<BikeRow>();
  return c.json({
    last_updated: ts,
    data: results.map((b) => ({
      bike_id: b.bike_id,
      lat: b.lat,
      lon: b.lon,
      is_reserved: !!b.is_reserved,
      is_disabled: !!b.is_disabled,
      vehicle_type_id: b.vehicle_type_id,
    })),
  });
});

bikes.get("/bikes/available", async (c) => {
  const city = resolveCity(c.req.query("city"));
  if (!city) return c.json({ detail: "Unknown city" }, 400);
  const ts = await latestBikeTimestamp(c.env.DB, city);
  if (!ts) return c.json({ detail: "No bike data available yet" }, 503);
  const { results } = await c.env.DB.prepare(
    "SELECT * FROM bike_snapshots WHERE city = ? AND timestamp = ? AND is_reserved = 0 AND is_disabled = 0"
  )
    .bind(city, ts)
    .all<BikeRow>();
  return c.json({
    last_updated: ts,
    available_bikes: results.map((b) => ({
      bike_id: b.bike_id,
      lat: b.lat,
      lon: b.lon,
      vehicle_type_id: b.vehicle_type_id,
    })),
  });
});

bikes.get("/ebikes", async (c) => {
  const city = resolveCity(c.req.query("city"));
  if (!city) return c.json({ detail: "Unknown city" }, 400);
  const ts = await latestBikeTimestamp(c.env.DB, city);
  if (!ts) return c.json({ detail: "No bike data available yet" }, 503);
  const { results } = await c.env.DB.prepare(
    "SELECT * FROM bike_snapshots WHERE city = ? AND timestamp = ? AND vehicle_type_id = 'bbe'"
  )
    .bind(city, ts)
    .all<BikeRow>();
  return c.json({
    last_updated: ts,
    ebikes: results.map((b) => ({
      bike_id: b.bike_id,
      lat: b.lat,
      lon: b.lon,
      is_reserved: !!b.is_reserved,
      is_disabled: !!b.is_disabled,
    })),
  });
});

export default bikes;
